econml/solutions/causal_analysis/_causal_analysis.py

```diff
@@ -1136,10 +1176,116 @@ def _heterogeneity_tree_string(self, Xtest, feature_index, *,
+                                                min_impurity_decrease=min_impurity_decrease,
+                                                alpha=alpha)
+
+        def hetero_data(tree, node_id):
+            return {'effect': _sanitize(tree.value[node_id])}
+        return _tree_interpreter_to_dict(intrp, feature_names, hetero_data)
+
+    def individualized_policy(self, Xtest, feature_index, *, n_rows=None, treatment_costs=0, alpha=0.1):
+        """
+        Get individualized treatment policy based on the learned model for a feature, sorted by the predicted effect.
+
+        Parameters
+        ----------
+        Xtest: array-like
+            Features
+        feature_index: int or string
+            Index of the feature to be considered as treatment
+        n_rows: int, optional
+            How many rows to return (all rows by default)
+        treatment_costs: array-like, default 0
+            Cost of treatment, as a scalar value or per-sample
+        alpha: float in [0, 1], default 0.1
+            Confidence level of the confidence intervals
+            A (1-alpha)*100% confidence interval is returned
+
+        Returns
+        -------
+        output: DataFrame
+            Dataframe containing recommended treatment, effect, confidence interval, sorted by effect
+        """
+        result = self._safe_result_index(Xtest, feature_index)
+
+        # get dataframe with all but selected column
+        orig_df = pd.DataFrame(Xtest, columns=self.feature_names_).drop(self.feature_names_[result.feature_index],
+                                                                        axis=1)
+
+        Xtest = result.X_transformer.transform(Xtest)
+        if Xtest.shape[1] == 0:
+            Xtest = None
+
+        effect = result.estimator.const_marginal_effect_inference(Xtest)
+        if result.feature_baseline is None:
+            # apply 10% of a typical treatment for this feature
+            effect.scale(result.treatment_value * 0.1)
+
+        effect.translate(-treatment_costs)
+
+        est = effect.point_estimate
+        est_lb = effect.conf_int(alpha)[0]
+        est_ub = effect.conf_int(alpha)[1]
+
+        if result.feature_baseline is None:
+            rec = np.empty(est.shape[0], dtype=object)
+            rec[est > 0] = "increase"
+            rec[est <= 0] = "decrease"
+            # set the effect bounds; for positive treatments these agree with
+            # the estimates; for negative treatments, we need to invert the interval
+            eff_lb, eff_ub = est_lb, est_ub
+            eff_lb[est <= 0], eff_ub[est <= 0] = -eff_ub[est <= 0], -eff_lb[est <= 0]
+            # the effect is now always positive since we decrease treatment when negative
+            eff = np.abs(est)
+        else:
+            # for discrete treatment, stack a zero result in front for control
+            zeros = np.zeros((est.shape[0], 1))
+            all_effs = np.hstack([zeros, est])
```

Comment:
As we discussed in our sync today, here the assumption is in discrete case, different treatments have equal cost.

Reply:
It turns out this need not be the case: treatment costs can be specified as an (n, d_t-1) array of costs that vary by individual and treatment.